File: 02_algorithm/programmers/Level3/섬 연결하기.py
def solution(n, costs):
    answer = 0
    costs.sort(key = lambda x: x[2])
    p = [x for x in range(n)]

    def find_set(x):
        if x != p[x]:
            p[x] = find_set(p[x])
        return p[x]

    MST = []
    cur = 0
    while len(MST) < n - 1:
        u, v, w = costs[cur]
        a = find_set(u)
        b = find_set(v)
        if a != b:
            p[b] = a
            MST.append([u, v, w])
        cur += 1
    
    for edge in MST:
        answer += edge[2]

    return answer


print(solution(4, [[0,1,1],[0,2,2],[1,2,5],[1,3,1],[2,3,8]])) # 4

File: 06_js/04_algorithm/programmers/Level3/섬 연결하기.js
function solution(n, costs) {
  let answer = 0;
  costs.sort((a, b) => a[2] - b[2]);
  let p = [];
  for (let i = 0; i < n; i++) {
    p.push(i);
  }
  
  const findSet = x => {
    if (x !== p[x]) {
      p[x] = findSet(p[x]);
    }
    return p[x];
  }
  
  let MST = [];
  let cur = 0;
  while (MST.length < n - 1) {
    const [u, v, w] = costs[cur];
    let a = findSet(u);
    let b = findSet(v);
    if (a !== b) {
      p[b] = a;
      MST.push([u, v, w]);
    }s
    cur += 1;
  }
  
  answer = MST.reduce((acc, curr) => acc += curr[2], 0);
  return answer;
}


console.log(solution(4, [[0,1,1],[0,2,2],[1,2,5],[1,3,1],[2,3,8]])); // 4